### PR TITLE
Hold reconfig lock while handling transaction

### PR DESCRIPTION
**Pull request by mystenmark**
This fixes a crash where we clear all pending locks at reconfig
(https://github.com/MystenLabs/sui/blob/dec33d0a303a7f83e778fcc5a136ab9776162e68/crates/sui-core/src/execution_cache/object_locks.rs#L121)
while trying to acquire locks for a transaction.

Without this fix, the node can reconfigure while we are trying to acquire locks for a transaction. If `clear_locks()` linked above is called while we are trying to call `clear_cached_locks` at https://github.com/MystenLabs/sui/blob/74d6d564970406e1b3191a07cf207af1ab6b3356/crates/sui-core/src/execution_cache/object_locks.rs#L245, we hit the panic at https://github.com/MystenLabs/sui/blob/74d6d564970406e1b3191a07cf207af1ab6b3356/crates/sui-core/src/execution_cache/object_locks.rs#L160

### CODE REVIEW

**Comment by lxfind** [comment]
Can you explain more? I don't fully understand how this could crash and why the PR fixes it

**Comment by mystenmark** [comment]
Added some more explanation to the description

**Comment by halfprice** [comment]
Looks like there is still compilation error?

**Comment by mystenmark** [comment]
> Looks like there is still compilation error?

indeed, thought it was a simple enough change to oneshot it but i was wrong

**Comment by mystenmark** [comment]
> > Looks like there is still compilation error?
> 
> indeed, thought it was a simple enough change to oneshot it but i was wrong

Oh, I used the wrong lock. Should work now.

File: crates/sui-core/src/authority.rs [hunk 1]
```diff
@@ -2887,6 +2880,11 @@ impl AuthorityState {
         }
     }
 
+    pub async fn execution_lock_for_signing(&self) -> ExecutionLockReadGuard {
+        // No need to check epoch here - current epoch is always correct for signing
```
Comment (lxfind) [inline crates/sui-core/src/authority.rs:2884]: Could you comment explain why signing also requires execution lock?

Comment (mystenmark) [reply crates/sui-core/src/authority.rs:2884]: Did you read the PR description? It's to prevent reconfiguration from happening while we are signing a tx.

Comment (lxfind) [reply crates/sui-core/src/authority.rs:2884]: Yes, but please also doc comment it.

**Review by lxfind — APPROVED** [review]

**Review by halfprice — APPROVED** [review]

File: crates/sui-core/src/authority.rs [hunk 2]
```diff
@@ -941,15 +941,8 @@ impl AuthorityState {
...
-        // between a validator signs a tx and a cert is formed, the validator
-        // could close the window.
-        if !epoch_store
-            .get_reconfig_state_read_lock_guard()
-            .should_accept_user_certs()
-        {
-            return Err(SuiError::ValidatorHaltedAtEpochEnd);
-        }
+        // Ensure that validator cannot reconfigure while we are signing the tx
+        let _execution_lock = self.execution_lock_for_signing().await;
```
Comment (aschran) [inline crates/sui-core/src/authority.rs:945]: can this lock be moved into `handle_transaction_impl` so it also works with #19401

**Review by aschran — CHANGES_REQUESTED** [review]

File: crates/sui-core/src/authority.rs (hunk 1 above)
Comment (mystenmark) [reply crates/sui-core/src/authority.rs:2884]: oh of course, I have no idea why i thought you meant "comment in the PR"

**Review by aschran — APPROVED** [review]

**Pull request merged**